Merge two Globbing Modules and Fix #5349

alt-romes opened the pull request:
Merge the two Globbing modules in cabal and cabal-install

We use the datatype representation from the globbing in cabal-install,
but preserve a standalone parser for globs present in cabal files, whose
specification is constrained by the cabal specification. The
implementations are merged taking the best parts of each.

and

Make sure sdist is robust to accidentally-listed directories

The refactor of the globbing modules re-uncovered this issue, and
required a prompt fix for the refactor not to break some things.

Fixes #5349



**Template Β: This PR does not modify `cabal` behaviour (documentation, tests, refactoring, etc.)**

Include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).

geekosaur commented on Cabal/src/Distribution/Simple/Glob.hs:
```diff
@@ -283,73 +483,111 @@ matchDirFileGlobWithDie verbosity rip version dir filepath = case parseFileGlob
+  -> FilePathGlobRel
+  -> IO [GlobResult FilePath]
+runDirFileGlob verbosity mspec rawRoot pat = do
+  -- The default data-dir is null. Our callers -should- b
```
```suggestion
  -- The default data-dir is null. Our callers -should- be
```

alt-romes replied:
Done

Kleidukos commented:
Ah, interesting failure: https://github.com/haskell/cabal/actions/runs/7745458515/job/21121495479?pr=9673#step:19:52

ffaf1 commented on Cabal/src/Distribution/PackageDescription/Check.hs:
```diff
@@ -170,7 +170,7 @@ checkPackageFilesGPD verbosity gpd root =
     checkPreIO =
       CheckPreDistributionOps
-        { runDirFileGlobM = \fp g -> runDirFileGlob verbosity (root </> fp) g
+        { runDirFileGlobM = \fp g -> runDirFileGlob verbosity (Just $ specVersion $ packageDescription gpd) (root </> fp) g
```
```suggestion
        { runDirFileGlobM = \fp g -> runDirFileGlob verbosity (Just . specVersion $ packageDescription gpd) (root </> fp) g
```

alt-romes replied:
Done

ffaf1 commented on Cabal-tests/tests/UnitTests/Distribution/Simple/Glob.hs:
```diff
@@ -101,13 +101,13 @@ testMatchesVersion version pat expected = do
     isEqual = (==) `on` (sort . fmap (fmap normalise))
     checkPure globPat = do
-      let actual = mapMaybe (fileGlobMatches globPat) sampleFileNames
+      let actual = mapMaybe (\p -> (p <$) <$> fileGlobMatches version globPat p) sampleFileNames
```
I am scratching my head a bit reading this.

alt-romes replied:
Fixed

ffaf1 commented on cabal-testsuite/PackageTests/Check/ConfiguredPackage/Paths/AbsolutePath/cabal.out:
```diff
@@ -1,6 +1,6 @@
 # cabal check
 These warnings may cause trouble when distributing the package:
-Warning: [no-glob-match] In 'extra-source-files': the pattern '/home/user/file' does not match any files.
+Warning: [glob-missing-dir] In 'extra-source-files': the pattern '/home/user/file' attempts to match files in the directory '/home/user', but there is no directory by that name.
```
This will fail if `Check/Warning.hs` is not changed. I am assuming this is still wip.

alt-romes replied:
Basically, we now check more consistently in some situations for whether the directory exists, resulting in "directory does not exist" error being triggered before "file does not match".

ffaf1 commented:
Also this is Template A.

alt-romes commented:
> Also this is Template A.

I don't think I have changed any behaviour, except for maybe the directory-not-exists error being reported instead of file-does-not-match, in some situations where the directory does not exist (in others, we would already report this correctly).
I've updated the tests accordingly.

andreabedini commented on Cabal/src/Distribution/Simple/Glob.hs:
```diff
@@ -283,73 +493,153 @@ matchDirFileGlobWithDie verbosity rip version dir filepath = case parseFileGlob
 isRecursiveInRoot _ = False
+
+-- | Check how the string matches the glob under this cabal version
+checkNameMatches :: CabalSpecVersion -> Glob -> String -> Maybe (GlobResult ())
+checkNameMatches spec glob candidate
```
This function needs unit tests

andreabedini commented on Cabal/src/Distribution/Simple/Glob.hs:
```diff
@@ -283,73 +493,153 @@ matchDirFileGlobWithDie verbosity rip version dir filepath = case parseFileGlob
+  -> FilePath
+  -> FilePathGlobRel
+  -> IO [GlobResult FilePath]
+runDirFileGlob verbosity mspec rawRoot pat = do
```
It is a big function, would it be hard to write unit tests?

andreabedini replied:
Thanks this makes it clearer (perhaps consider adding this as a comment). What you just said suggests changing `Glob` to represent the semantic of the globbing rather than the only the syntax (I assume they diverge in few specific cases); and adding a normalization function `normaliseForCabalSpec :: CabalSpecVersion -> Glob -> Glob`? (maybe even distinguishing the two `Glob` at the type level). Then`runDirFileGlob` is a mere interpreted independent from the context.

I am looking again and I think the code would be more readable if done in this way. E.g. ```
  isFile <- maybe (return True) (const $ doesFileExist (root </> dir </> s)) mspec
```
is a bit of a lie right? it actually mean `isFileOrMSpecIsNull`.

I really think there is a better way to write this.

andreabedini commented on Cabal/src/Distribution/Simple/Glob.hs:
```diff
@@ -283,73 +493,153 @@ matchDirFileGlobWithDie verbosity rip version dir filepath = case parseFileGlob
+-- | How/does the glob match the given filepath, according to the cabal version?
+-- Since this is pure, we don't make a distinction between matching on
+-- directories or files (i.e. this function won't return 'GlobMatchesDirectory')
+fileGlobMatches :: CabalSpecVersion -> FilePathGlobRel -> FilePath -> Maybe (GlobResult ())
+fileGlobMatches version g path = go g (splitDirectories path)
```
This function needs unit tests

andreabedini reviewed (changes requested):
Generally ok with it. It's a good cleanup. I think we need unit tests for a few functions (at least the pure ones, which are trivial to test).

alt-romes commented:
> Generally ok with it. It's a good cleanup. I think we need unit tests for a few functions (at least the pure ones, which are trivial to test).

We already do! Globbing is extensively tested in all of its behavior, both for the Client.Globing and Simple.Globbing, and I've been making all of these tests pass.

alt-romes commented on Cabal/src/Distribution/PackageDescription/Check.hs:
```diff
@@ -854,12 +854,14 @@ checkGlobResult title fp rs = dirCheck ++ catMaybes (map getWarning rs)
+    -- (currently) support disjunction, (ROMES:TODO: We do now...)
+    -- that will always mean there are
     -- no matches. The no matches error in this case is strictly less
     -- informative than the missing directory error.
```
Done.

alt-romes commented:
Here are some pointers to our current unit tests on globbing:

We tests both pre-2.4 and post-2.4 spec version by matching against these files

https://github.com/haskell/cabal/blob/589fa5603a7357aec8a8752fa5e6f140db8786fb/Cabal-tests/tests/UnitTests/Distribution/Simple/Glob.hs#L19-L43

and expecting these results

https://github.com/haskell/cabal/blob/589fa5603a7357aec8a8752fa5e6f140db8786fb/Cabal-tests/tests/UnitTests/Distribution/Simple/Glob.hs#L52-L81

More specifically, we test both the pure version of the globbing `fileGlobMatches` (which is basically a small wrapper calling`checkNameMatches`), and the IO version `runDirFileGlob` which checks the existing directories and files in tree:
https://github.com/haskell/cabal/blob/589fa5603a7357aec8a8752fa5e6f140db8786fb/Cabal-tests/tests/UnitTests/Distribution/Simple/Glob.hs#L103-L112

andreabedini commented:
> We already do! Globbing is extensively tested in all of its behavior, both for the Client.Globing and Simple.Globbing, and I've been making all of these tests pass.

So all this is tested at the top, ok. It can be useful to have test also for helper functions but no worries. I marked those comments as resolved.

andreabedini commented on Cabal/src/Distribution/Simple/Glob.hs:
```diff
@@ -283,73 +493,153 @@ matchDirFileGlobWithDie verbosity rip version dir filepath = case parseFileGlob
+              -- when a cabal spec version is passed as an argument), we
+              -- disallow matching a @GlobFile@ against a directory, preferring
+              -- @GlobDir dir GlobDirTrailing@ to specify a directory match.
+              isFile <- maybe (return True) (const $ doesFileExist (root </> dir </> s)) mspec
```
This tests if the file exists and is not a directory, but later we might return `GlobMatchesDirectory` ... without checking if the directory exists? Do I read this right?

alt-romes replied:
We are sure that the directory is, because we only run this predicate on the results of
```
entries <- getDirectoryContents (root </> dir)
```
which are definitely things that exist.

andreabedini commented:
@alt-romes thanks for walking me through this today. I left the changes we did together in https://github.com/alt-romes/cabal/pull/1, feel free to take what you like.

sheaf commented on Cabal/src/Distribution/Simple/Glob.hs:
```diff
@@ -14,59 +17,331 @@
+-- If it is trivial in this sense then the result is the equivalent constant
+-- 'FilePath'. On the other hand if it is not trivial (so could in principle
+-- match more than one file) then the result is @Nothing@.
+isTrivialFilePathGlob :: FilePathGlob -> Maybe FilePath
```
I've moved the function back to `cabal-install`. It's meant to check whether the glob is equivalent to a simple filepath (that uses no globbing syntax). Seems self-explanatory enough, but I don't really understand why e.g. we return `Nothing` in the `FiltePathHomeDir` case. I'm going to leave it as-is for now as this just preserves the status quo.

sheaf commented:
I moved back the `cabal-install`-specific parts to `Distribution.Client.Glob`. This means we keep sharing the functionality, but don't end up polluting the `Cabal` library with these details.

sheaf commented:
> @alt-romes thanks for walking me through this today. I left the changes we did together in [alt-romes#1](https://github.com/alt-romes/cabal/pull/1), feel free to take what you like.

I didn't apply these changes, as they regressed `T6125` for me.

alt-romes commented:
@andreabedini light ping. Sam has improved the MR and addressed the rest of the review.
Shall we merge it?

fgaz reviewed (changes requested):
[This comment](https://github.com/haskell/cabal/pull/9673#issuecomment-1922221173) still has to be addressed

> Also this is Template A.

alt-romes commented:
> [This comment](https://github.com/haskell/cabal/pull/9673#issuecomment-1922221173) still has to be addressed
> 
> > Also this is Template A.

I replied here:

> I don't think I have changed any behaviour, except for maybe the directory-not-exists error being reported instead of file-does-not-match, in some situations where the directory does not exist (in others, we would already report this correctly).
I've updated the tests accordingly.

alt-romes commented:
@fgaz what specifically do you think we're missing?

fgaz commented:
Oh, sorry, I missed that comment.

If the Cabal API changed pheraps add a changelog entry and/or `@since` annotations

fgaz commented on Cabal/src/Distribution/Simple/Glob.hs:
```diff
@@ -14,59 +14,247 @@
-  , isRecursiveInRoot
-  , Glob
-  ) where
+module Distribution.Simple.Glob where
```
Can we keep the export list? We probably want to avoid exporting implementation details if possible

sheaf replied:
Done.

andreabedini commented:
I'm ok but please do add a changelog entry and @since annotations and consider explicitly listing the intended exports of Distribution.Simple.Glob.

ffaf1 commented on Cabal/src/Distribution/PackageDescription/Check/Monad.hs:
```diff
@@ -48,7 +48,7 @@ import Distribution.CabalSpecVersion (CabalSpecVersion)
 import Distribution.PackageDescription.Check.Warning
 import Distribution.Simple.BuildToolDepends (desugarBuildToolSimple)
-import Distribution.Simple.Glob (Glob, GlobResult)
+import Distribution.Simple.Glob
```
Is thre a good reason not to have explicit imports?

sheaf replied:
Sorry, I put them back to what they were before.

ffaf1 commented on Cabal/src/Distribution/PackageDescription/Check.hs:
```diff
@@ -871,6 +872,7 @@ checkGlobResult title fp rs = dirCheck ++ catMaybes (map getWarning rs)
       Just $ PackageDistSuspiciousWarn (GlobExactMatch title fp file)
     getWarning (GlobMissingDirectory dir) =
       Just $ PackageDistSuspiciousWarn (GlobNoDir title fp dir)
+    getWarning (GlobMatchesDirectory _) = Nothing -- is handled elsewhere if relevant, it is not necessarily a problem
```
If possible move comment up a line, capital letter, full stop.

sheaf commented:
The only changes in the user-facing API are:
 
  - addition of the `GlobMatchesDirectory` constructor of `GlobResult`,
  - new function `globMatches`.

I made a separate `Distribution.Simple.Glob.Internal` module for the internal implementation details, in order to preserve the user-facing API of `Distribution.Simple.Glob`. Note that `Distribution.Simple.Glob.Internal` is nonetheless an exposed module of the `Cabal` library, because `cabal-install` file monitoring needs access to the internals.

sheaf commented:
> I'm ok but please do add a changelog entry and @SInCE annotations and consider explicitly listing the intended exports of Distribution.Simple.Glob.

Done.
